---
title: "Integration of RNAseq Analysis tools with RNAseqEasy"
author: Salvador Torres Montilla
date: "Last edited `r format(Sys.time(), '%d %B, %Y')`"
abstract: >
  RNAseq analysis has become a common practice in most molecular biology
  laboratories in the world. Several tools have been developed to achieve 
  differential expression analysis, functional enrichment, gene co-expression 
  or data visualization among others, from RNAseq data. However, many of them
  require independent analysis and, in many ocasions, relatively high
  programming skills. The package RNAseqEasy provides an integration of some of
  this tools into a single user-friendly workflow, and tries to simplify it to 
  require only basic notions in R programming. It is mainly based on DESeq2 
  [@DESeq2], topGO [@topGO] and WGCNA [@WGCNA] packages to carry out PCA
  exploratory plots, differential expression analysis, gene ontology enrichment 
  analysis and gene-coexpression from Salmon [@patro2017] quantified data. This 
  vignette explains the use of the package and demonstrates typical workflows.
  
  
  RNAseqEasy package version: `r packageVersion("RNAseqEasy")`
output: 
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    fig_width: 5
vignette: >
  %\VignetteIndexEntry{Integration of RNAseq Analysis tools with RNAseqEasy}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
bibliography: references.bib
csl: nature.csl
---

```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(tidy = FALSE,
                      cache = FALSE,
                      dev = "png",
                      message = FALSE, error = FALSE, warning = TRUE)
temp_output_dir <- tempdir()
library(RNAseqEasy)
```

## Input data

The RNAseqEasy package is thought to be used from Salmon mapped and quantified data. The first required input is the path [1] of the directory where `quant.sf` samples are saved will be the input. The other input necessary for the analysis is a data frame [2] that includes the name of the samples and the information describing them (i.e. variables or conditions defining the experiment).

### Salmon data

Salmon [@patro2017] is a widely used software for quantifying transcript abundance. A well detailed tutorial can be found in their [official webpage](https://combine-lab.github.io/salmon/getting_started/). It is a very fast tool for transcript quantification directly from `fastq.gz`files. The only requirement is the creation of an index of the **transcriptome** of the organism we are working with (do not use the genome!!!), also explained in their tutorial.

As a result, we will obtain a folder [1] containing one folder per analyzed sample with its names. In each folder, the main output fail will be named `quant.sf`, which contains the name of each transcript and their abundance in Transcripts Per Million (TPM), among other information (length, effective length and number of reads). The path to this output folder will be input for our analysis [1].

```{r zenodo}
# URL from Zenodo
data_url <- "https://zenodo.org/records/15800134/files/GSE275561.zip?download=1"

# Temporal directory to download data
output_dir <- file.path(tempdir(), "GSE275561")
dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

zip_file <- file.path(tempdir(), "GSE275561.zip")

# Download zip file with Salmon quantified data from Zenodo
download.file(url = data_url, destfile = zip_file, mode = "wb")

# Unzip compressed file
unzip(zip_file, exdir = output_dir)

list.files(output_dir)
```

So, we set the "02_Salmon" folder as `sampleDir`, which will direct the analysis to the directory to retrieve quantified data for each sample.

```{r sampleDir}
sampleDir <- file.path(output_dir, "02_Salmon")
```

### Sample table

For each experiment analysis, it is required to generate a data frame [2] (we call it `sample_table`) which correlates the name of each sequenced sample with the variables describing it in the experiment. So, it will include a `Sample` column including sample names, and an extra column for each variable of factor that were included in the experiment design. In our example, we include 12 different samples (Wendy1 to Wendy12), and there were 3 different variables describing them:

-   `Genotype`. Samples were divided in two different genotypes: 'Tak1' (wild-type organisms) and 'gh3a' (CRISPR-Cas9 mutants).
-   `Treatment`. Two different treatments were applied to samples: 'Mock' or 'OPDA' (exposure to a high concentration of a plant stress hormone from jasmonic acid family).
-   `Replicate`. Each category of samples included 3 independent biological replicates.

```{r sample_data}
# Data frame with 'Sample', 'Genotype', 'Treatment and 'Replicate' information

sample_table <- data.frame(
  Sample = paste0("Wendy", seq(1,12)),
  Genotype = rep(c("Tak1", "gh3a"), each = 6),
  Treatment = rep(c("Mock", "OPDA"), each = 3),
  Replicate = seq(1,3)
)

sample_table
```

In differential expression analysis, comparisons are performed against a reference level. By default, R set levels based on alphabetical order for each variable. We can set the specific comparison we want to perform in each analysis, but it is a good practice to set levels order as we want to before we continue. We transform each variable column into factor and establish the level order. In this example, "TaK-1" is the reference genotype, and "Mock" is the reference treatment.

```{r factors}
# Convert 'Genotype' into factor, setting "Tak1" as reference
sample_table$Genotype <- factor(sample_table$Genotype,
                                levels = c("Tak1", "gh3a"))

# Convert 'Treatment' into factor, setting "Mock" as reference
sample_table$Treatment <- factor(sample_table$Treatment, levels = c("Mock", "OPDA"))

str(sample_table)
```

### Functional annotation

Everything else required for the analysis is not specific of the experiment you are performing, but depends on the organism you are working with. There are two different files that we will need to import for the analysis:

-   A data frame that correlates each transcript name with its corresponding gene name.
-   A file including functional annotation, i.e., correlating each gene/transcript to GO terms.

In this example, we are working with samples extracted from the liverwort *Marchantia polymorpha*, a model plant widely use to study how ancient or conserved biological processes are in plant evolution. To get the reference transcriptome, we download it from [marchantia.info](https://marchantia.info/), the Genome Database for *Marchantia polymorpha*. The `get.fasta.name()` function from phylotools [@revell2024] package allows us to obtain the transcript names from the fasta file. In *Marchantia*, gene names and transcript share the same nomenclature, only adding a dot and a number to the different transcripts of a gene. To correlate them, we create a data frame including a column called `Name` containing transcript names, and based on that we create a new column called `GENEID` removing last 2 characters (using `str_sub()` function from string package).

```{r mRNA_fasta, warning=FALSE}
library(phylotools)
library(tidyverse)
URL <- gzcon(url(paste("https://marchantia.info/download/MpTak_v6.1/", "MpTak_v6.1r1.mrna.fasta.gz", sep="")))
txt <- readLines(URL)
Marchantia7_Transcripts <- get.fasta.name(textConnection(txt), clean_name = FALSE)
head(Marchantia7_Transcripts)

Marchantia7_tx2gene <- data.frame(Name = Marchantia7_Transcripts) %>%
  tidyr::separate(Name, sep = " ", into = c("TXNAME", "CDS")) %>%
  dplyr::mutate(GENEID = stringr::str_sub(TXNAME, 1, -3)) %>%
  dplyr::select(-CDS)

head(Marchantia7_tx2gene, 10)
```

For this example, functional annotation of Gene Ontologies (GOs) for *Marchantia* genes can be downloaded from the package in a csv file. Many organisms reference databases include this kind of files. If you cannot find them, there are tools and tutorial to annotate your reference genome.

```{r FunctionalDB}
library(RNAseqEasy)

GO_data <- system.file("extdata",
                      "Mpo6.1_GO_db_GeneGO.csv",
                      package="RNAseqEasy", mustWork=TRUE)
Mpo_GO_GOSLIM <- read.csv(GO_data, sep = "\t")
head(Mpo_GO_GOSLIM)
```

This annotation file is a two column data frame correlating transcript names and GO terms (one assotiation per row). For topGO GO enrichement analyses, we will need to convert it into a list containing transcript as keys and all their annotated GO terms as values. The `load_topGO_db()` function included in this package helps us in this purpose.

```{r loadGO}
geneID2GO <- load_topGO_db(Mpo_GO_GOSLIM, "Transcript", "GO")
head(geneID2GO)
```

To perform GO semantic clustering, the analysis is based on rrvgo [@sayols2023] and GOSemSim [@yu2020] packages. The available species for this analysis are included in [Bioconductor webpage](https://bioconductor.org/packages/release/BiocViews.html#___OrgDb). The only plant organism included is *Arabidopsis thaliana*, so we will use its code (org.At.tair.db) for this analysis. We can save this data in an object, so it will save computing time in the following analysis. There are three classes of GO terms: Biological Processes (BP), Molecular Functions (MF) and Cell Components (CC). In this example, we will only be focused on biological processes.

```{r At_tair, warning=FALSE}
save <- GOSemSim::godata("org.At.tair.db", ont="BP")
```

## Differential gene expression analysis

At this point, we already have everything we need to perform our RNA-seq analysis: a directory with our quantified samples (`sampleDir`), a data frame with the experiment metadata (`sample_table`), a data frame correlating gene names and transcript names for our organism (`Marchantia7_tx2gene`), an annotation list correlating transcript names and their annotated GO terms (`geneID2GO`) and the model organism reference data for the semantic clustering (`save`).

We want to obtain differential expressed genes in **Tak-1 genotype in OPDA treatment compared to Tak-1 genotype in Mock conditions**. To organize our results, we will create a folder to save all differential expression analyses and we will call it "DESeq2", and inside we will create another folder to save specifically this comparison, and we will call it "Tak1_OPDA".

```{r ouputDirDESeq2}
dir.create(file.path(output_dir, "DESeq2"))
dir.create(file.path(output_dir, "DESeq2", "Tak1_OPDA"))
output_Dir_DESeq2 <- file.path(output_dir, "DESeq2", "Tak1_OPDA")
```

To run the analysis, we should install [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) and [topGO](https://bioconductor.org/packages/release/bioc/html/topGO.html) packages from Bioconductor and load them in our R session.

Now, we are ready to run our differential expression analysis. For that, we will use `DESeq2_simple` function. We set all required parameters that we previously defined (`output_path`, `sampleDir`, `sample_table`, `tx2gene`, `geneID2GO`, `orgdb` and `semdata`). But, still, there are some parameters that we have to define for this analysis:

-   **Name**: We have to include a string to name all files that are going to be generated. In this example, we use the same as the folder where they will be saved ("Tak1_OPDA").
-   **Variable**: Vector of variables or factors that are affecting gene expression in the experiment.
-   **Design**: Design formula indicating the variables which will be used in the modelling, i.e., which variables we expect that are having an effect in gene expression in the analysis. In this example, we expect that "Genotype" itself (there will be genes which expression will be expected to be affected only by genotype in any condition), "Treatment" itself (there will be genes which expression will be expected to be affected only by treatment in any genotype), and the interaction "Genotype+Treatment" ((there will be genes which expression will be expected to be affected differently by treatment in a genotype compared to the other)), will be affecting gene expression. Therefore, our design formula will be "Genotype + Treatment + Genotype:Treatment". (For more information about design, read the [DESeq2 vignette](https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)).
-   **Contrast**: The specific contrast we want to perform to test differential expression. If we do not specify any contrast, the function will stop and there will be printed all possible contrast we can use in the experimental design we are using. There are two different ways to define the contrast we want to:
    -   Condition A - Condition B.
    -   list(c("Contrast")).

For more information about contrasts, read the [DESeq2 vignette](https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) and check this [tutorial](https://github.com/tavareshugo/tutorial_DESeq2_contrasts).

```{r SelectContrast, error = TRUE, warning=FALSE}
library(DESeq2)
library(topGO)

DESeq2_simple(
    output_path = output_Dir_DESeq2,
    sampleDir = sampleDir,
    sample_table = sample_table,
    tx2gene = Marchantia7_tx2gene,
    geneID2GO = geneID2GO,
    orgdb = "org.At.tair.db",
    semdata = save,
    Name = "Tak1_OPDA",
    Variable = c("Genotype", "Treatment"),
    Design = "Genotype + Treatment + Genotype:Treatment"
  )
```

After running `DESeq2_simple` with no Contrast parameter, there is an error telling us we have to choose one of the previous contrasts printed in the console. In this example, we want to compare Tak-1 genotype in OPDA treatment samples vs. Tak-1 genotype in Mock conditions. There are two different ways we can set this contrast:

-   Tak1_OPDA - Tak1_Mock.
-   list(c("Treatment_OPDA_vs_Mock")). Since we established Tak-1 as our reference genotype (first factor), the treatment OPDA vs Mock contrast will get this comparison in the reference genotype, i.e., Tak-1.

```{r DESeq2_simple, warning=FALSE}
result <- DESeq2_simple(
    output_path = output_Dir_DESeq2,
    sampleDir = sampleDir,
    sample_table = sample_table,
    Include = NULL,
    Exclude = NULL,
    tx2gene = Marchantia7_tx2gene,
    Variable = c("Genotype", "Treatment"),
    Design = "Genotype + Treatment + Genotype:Treatment",
    Group = "NO",
    Name = "Tak1_OPDA",
    Contrast = list(c("Treatment_OPDA_vs_Mock")),
    Reduced = FALSE,
    log2FCtopGO = 1,
    geneID2GO = geneID2GO,
    ontology = "BP",
    plot_similarity = TRUE,
    orgdb = "org.At.tair.db",
    semdata = save
  )
```

If we want to perform other possible analyses, these would be the contrasct we would have to use:

-   **OPDA vs Mock in *gh3a* genotype**: gh3a_OPDA - gh3a_Mock, or list(c("Treatment_OPDA_vs_Mock", "Genotypegh3a.TreatmentOPDA")).

-   ***gh3a*** **vs Tak-1 in Mock conditions**: gh3a_Mock - Tak1_Mock, or list(c("Genotype_gh3a_vs_Tak1")).

-   ***gh3a*** **vs Tak-1 in OPDA conditions**: gh3a_OPDA - Tak1_OPDA, or list(c("Genotype_gh3a_vs_Tak1", "Genotypegh3a.TreatmentOPDA")).

-   **Genes affected by the Genotype:Treatment interaction**: (gh3a_OPDA - gh3a_Mock) - (Tak1_OPDA - Tak1_Mock), or list(c("Genotypegh3a.TreatmentOPDA")). This are genes that will be responding to OPDA treatment in the *gh3a* genotype in a different way that the expected from the response in Tak-1 genotype.

```{r PCA, echo=FALSE, fig.cap="PCA plot of PC1 vs PC2, using 500 top genes"}
ruta_PCA_generado <- file.path(output_Dir_DESeq2, "PCA_Tak1_OPDA.png")
knitr::include_graphics(ruta_PCA_generado)
```

```{r scatterplot, echo=FALSE, fig.cap="Semantic cluster of Gene Ontologies enriched in Up-regulated genes"}
ruta_pdf_generado <- file.path(output_Dir_DESeq2, "topGO", "topGO_Tak1_OPDA_Up_Scatterplot.pdf")
rutas_png_generadas <- pdftools::pdf_convert(
  pdf = ruta_pdf_generado,
  format = "png",
  dpi = 150 # Ajusta la calidad de la imagen según necesites
)
knitr::include_graphics(rutas_png_generadas[1])
```
